From 7956fb1e11494b64a3dbf48cd5f6900cead669bc Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Sun, 20 Mar 2022 12:02:07 +0300
Subject: [PATCH 038/151] arm64: dts: renesas: r8a779f0-vc4: dt overlay based
 phy mux setup

Modify r8a779f0-vc4.dts:
- add nodes for two PHYs per TSN,
- mark all PHYs disabled,
- remove phy-handle links for TSNs.

Introduce dt overlays that set up PHY mux. Each overlay:
- configures needed gpio hogs,
- configures phy-handle link,
- enables selected PHY.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 arch/arm64/boot/dts/renesas/Makefile          | 11 ++++-
 .../dts/renesas/r8a779f0-vc4-tsn0-phy-1g.dtso | 30 ++++++++++++
 .../renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso    | 48 +++++++++++++++++++
 .../dts/renesas/r8a779f0-vc4-tsn1-phy-1g.dtso | 21 ++++++++
 .../renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso    | 39 +++++++++++++++
 .../renesas/r8a779f0-vc4-tsn2-phy-1g-rh.dtso  | 21 ++++++++
 .../dts/renesas/r8a779f0-vc4-tsn2-phy-1g.dtso | 21 ++++++++
 arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts  | 33 +++++++++----
 8 files changed, 214 insertions(+), 10 deletions(-)
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-1g.dtso
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-1g.dtso
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g-rh.dtso
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g.dtso

diff --git a/arch/arm64/boot/dts/renesas/Makefile b/arch/arm64/boot/dts/renesas/Makefile
index 25936ebca4f4..afa2a477b2da 100644
--- a/arch/arm64/boot/dts/renesas/Makefile
+++ b/arch/arm64/boot/dts/renesas/Makefile
@@ -97,7 +97,7 @@ dtb-$(CONFIG_ARCH_R8A77980) += r8a77980-es2-condor.dtb
 dtb-$(CONFIG_ARCH_R8A77980) += r8a77980-es2-v3hsk.dtb
 
 # overlays
-# DTC_FLAGS += -@
+DTC_FLAGS += -@
 # ulcb-kf-dtbo := ulcb-kf-cn11.dtbo ulcb-kf-most.dtbo ulcb-kf-sd3.dtbo \
 #		ulcb-kf-cn10-pca.dtbo ulcb-kf-cn11-pca.dtbo \
 #		ulcb-kf-cn10-gmsl2.dtbo ulcb-kf-cn11-gmsl2.dtbo \
@@ -113,3 +113,12 @@ dtb-$(CONFIG_ARCH_R8A77980) += r8a77980-es2-v3hsk.dtb
 # dtb-$(CONFIG_ARCH_R8A77965) += $(ulcb-kf-dtbo)
 # dtb-$(CONFIG_ARCH_R8A77970) += $(vb-dtbo)
 # dtb-$(CONFIG_ARCH_R8A77980) += $(vb-dtbo)
+
+r8a779f0-vc4-dtbo :=	r8a779f0-vc4-tsn0-phy-1g.dtbo \
+			r8a779f0-vc4-tsn0-phy-2g5.dtbo \
+			r8a779f0-vc4-tsn1-phy-1g.dtbo \
+			r8a779f0-vc4-tsn1-phy-2g5.dtbo \
+			r8a779f0-vc4-tsn2-phy-1g.dtbo \
+			r8a779f0-vc4-tsn2-phy-1g-rh.dtbo
+
+dtb-$(CONFIG_ARCH_R8A779F0) += $(r8a779f0-vc4-dtbo)
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-1g.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-1g.dtso
new file mode 100644
index 000000000000..0fde7e6d2f74
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-1g.dtso
@@ -0,0 +1,30 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&gpio3 {
+	tsn0-pre-mux {
+		gpio-hog;
+		gpios = <8 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-PRE-MUX";
+	};
+};
+
+&gpio4 {
+	tsn0-mux {
+		gpio-hog;
+		gpios = <6 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-MUX";
+	};
+};
+
+&tsn0 {
+	phy-handle = <&tsn0_phy_1g>;
+};
+
+&tsn0_phy_1g {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso
new file mode 100644
index 000000000000..3cd077be73f5
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn0-phy-2g5.dtso
@@ -0,0 +1,48 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&gpio3 {
+	tsn0-pre-mux {
+		gpio-hog;
+		gpios = <8 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-PRE-MUX";
+	};
+};
+
+&gpio4 {
+	tsn0-mux {
+		gpio-hog;
+		gpios = <6 GPIO_ACTIVE_HIGH>;
+		output-low;
+		line-name = "TSN0-MUX";
+	};
+};
+
+&gpio3 {
+	tsn0-no-sleep {
+		gpio-hog;
+		gpios = <15 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-NO-SLEEP";
+	};
+};
+
+&gpio6 {
+	tsn0-no-wake {
+		gpio-hog;
+		gpios = <22 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-NO-WAKE";
+	};
+};
+
+&tsn0 {
+	phy-handle = <&tsn0_phy_2g5>;
+};
+
+&tsn0_phy_2g5 {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-1g.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-1g.dtso
new file mode 100644
index 000000000000..45db0ad299d9
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-1g.dtso
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&gpio4 {
+	tsn1-mux {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN1-MUX";
+	};
+};
+
+&tsn1 {
+	phy-handle = <&tsn1_phy_1g>;
+};
+
+&tsn1_phy_1g {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso
new file mode 100644
index 000000000000..aba41c950217
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn1-phy-2g5.dtso
@@ -0,0 +1,39 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&gpio4 {
+	tsn1-mux {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		output-low;
+		line-name = "TSN1-MUX";
+	};
+};
+
+&gpio3 {
+	tsn1-no-sleep {
+		gpio-hog;
+		gpios = <12 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN1-NO-SLEEP";
+	};
+};
+
+&gpio6 {
+	tsn1-no-wake {
+		gpio-hog;
+		gpios = <21 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN1-NO-WAKE";
+	};
+};
+
+&tsn1 {
+	phy-handle = <&tsn1_phy_2g5>;
+};
+
+&tsn1_phy_2g5 {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g-rh.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g-rh.dtso
new file mode 100644
index 000000000000..4bd3434b7000
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g-rh.dtso
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&gpio3 {
+	tsn2-mux {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		output-low;
+		line-name = "TSN2-MUX";
+	};
+};
+
+&tsn2 {
+	phy-handle = <&tsn2_phy_1g_rh>;
+};
+
+&tsn2_phy_1g_rh {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g.dtso b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g.dtso
new file mode 100644
index 000000000000..141e40608b86
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-tsn2-phy-1g.dtso
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+&gpio3 {
+	tsn2-mux {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN2-MUX";
+	};
+};
+
+&tsn2 {
+	phy-handle = <&tsn2_phy_1g>;
+};
+
+&tsn2_phy_1g {
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
index 98c2346245e1..aabf78f38167 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
@@ -269,28 +269,43 @@
 	status = "okay";
 
 	ports {
-		port@0 {
+		tsn0: port@0 {
 			phy-mode = "sgmii";
-			phy-handle = <&etha0>;
-			etha0: ethernet-phy@0 {
+			/* phy-handle set via plugin */
+			tsn0_phy_2g5: ethernet-phy@0 {
+				reg = <0>;
+				status = "disabled";
+			};
+			tsn0_phy_1g: ethernet-phy@3 {
 				reg = <3>;
 				compatible = "ethernet-phy-id002B.0218";
+				status = "disabled";
 			};
 		};
-		port@1 {
+		tsn1: port@1 {
 			phy-mode = "sgmii";
-			phy-handle = <&etha1>;
-			etha1: ethernet-phy@1 {
+			/* phy-handle set via plugin */
+			tsn1_phy_2g5: ethernet-phy@0 {
+				reg = <0>;
+				status = "disabled";
+			};
+			tsn1_phy_1g: ethernet-phy@3 {
 				reg = <3>;
 				compatible = "ethernet-phy-id002B.0218";
+				status = "disabled";
 			};
 		};
-		port@2 {
+		tsn2: port@2 {
 			phy-mode = "sgmii";
-			phy-handle = <&etha2>;
-			etha2: ethernet-phy@2 {
+			/* phy-handle set via plugin */
+			tsn2_phy_1g_rh: ethernet-phy@1 {
+				reg = <1>;
+				status = "disabled";
+			};
+			tsn2_phy_1g: ethernet-phy@3 {
 				reg = <3>;
 				compatible = "ethernet-phy-id002B.0218";
+				status = "disabled";
 			};
 		};
 	};
-- 
2.17.1

